accepted/00000-modular-repos-with-clm.md

```diff
@@ -0,0 +1,180 @@
+Users enable/disable streams so that there is only a single stream of a module is visible to the package manager at once.
+
+Modular repositories are defined in a new type of metadata called `modulemd` in a file named `modules.yaml`. This metadata has to be interpreted properly to manage these repositories.
+Unless this metadata is handled, managing modular repositories causes problems with package operations, specifically, suggesting incorrect package updates in the UI.
+
+Currently, Uyuni has only the most basic support for modularity by displaying the following message in the package management UI:
+```
+At least one of the channels this system is subscribed to contains modules. If you have activated modules on this system, please refrain from using Spacewalk for package operations. Instead, perform all package actions from the client using dnf.
+```
+
+The goal is to handle modular repositories in a way that the clients can be still managed through the Uyuni UI.
+
+# Detailed design
+[design]: #detailed-design
+
+## Overview
+
+Instead of implementing the modularity logic inside the core of Uyuni, this document describes a design to simulate the modularity behavior by building regular repositories out of modular repositories by filtering out packages from unselected streams.
+The resulting repository is a "flat" repository with no modularity, so it can be managed with Uyuni's existing package operations logic.
+
+Conceptually, a module can be in one of the following three states at any time:
+ - Enabled
+ - Disabled
+ - Neither enabled nor disabled
+
+When a module is neither enabled nor disabled, all the alternative streams and contexts of a stream are available to the package manager. This state confuses Uyuni when comparing the package versions.
+
+The key to building a regular repository with no conflicting packages is to simulate the modules in a way that every single module is either "enabled" or "disabled".
+
+Therefore, the high-level algorithm is proposed as:
+ 1. Let the users pick desired streams so that every module is either "enabled" or "disabled"
+ 2. Compile a blacklist of RPM artifacts from the "disabled" modules
+ 3. Generate a "regular" repository, filtering out the blacklisted RPMs
+
+The design utilizes content lifecycle management filters and repository building capabilities to implement the algorithm above.
+
+## Components
+
+This section summarizes the different components involved in the design.
+
+### Backend
+The Java backend is responsible for filtering out unselected module streams and building the repository with the desired packages through CLM.
+
+### API
+A Python API parses and interprets the module metadata, resolves modular dependencies and translates specified modules and streams to an actual package list.
+The API utilizes the Python port of [libmodulemd](https://github.com/fedora-modularity/libmodulemd), which is used for structural parsing of `modules.yaml` files.
```

Comment:
Libmodulemd is only available in PackageHub and only for Python 3. 

I assume we will distribute in the SUSE Manager channels for SLE15 SP1. What about SP2? Still in our channels?

Is it OK to have it only for Python 3?

Reply:
It is only required for SUSE Manager server, so Python 3 will be sufficient.